schnorr_fun/src/frost_backup.rs

```diff
@@ -0,0 +1,198 @@
+use secp256kfun::{
+    digest::{generic_array::typenum::U32, Digest},
+    marker::Secret,
+    Point, Scalar,
+};
+
+/// Secret shares can have a small index up to 32, or index at a specific scalar.
+#[derive(Debug, Clone, Eq, PartialEq)]
+pub enum BackupShareIndex {
+    /// a secret share at a small index up to 32
+    SmallIndex(usize),
+    /// a secret share at a specific scalar index
+    Scalar(Scalar),
+}
+
+/// An error encountered when encoding a Frostsnap backup.
+#[derive(Debug, Copy, Clone)]
+pub enum FrostBackupError {
+    /// Generic error from bech32 library
+    Bech32Error(bech32::Error),
+    /// Threshold was zero
+    ThresholdCantBeZero,
+    /// Threshold was greater than 32.
+    ThresholdTooBig,
+    /// Tried to decode a bech32 variant that was not bech32m
+    WrongBech32Variant(bech32::Variant),
+    /// Decoded secret share is not a valid secp256k1 scalar
+    InvalidSecretShareScalar,
+    /// Tried to decode backup with unknown prefix
+    InvalidHumanReadablePrefix,
+    /// The share index data length doesn't match expected for SmallIndex nor Scalar
+    UnknownShareIndexType,
+    /// Decoded share index is not a valid secp256k1 scalar
+    InvalidShareIndexScalar,
+}
+
+/// Create an identifier that's used to determine compatibility of FROST secret shares.
+/// The first 4 bech32 chars from a hash of the polynomial coefficients.
+/// Collision expected once in (32)^4 = 2^20.
+pub fn polynomial_identifier<H: Default + Digest<OutputSize = U32>>(
+    polynomial: Vec<Point>,
+) -> [u5; 4] {
+    let mut hash = H::default();
+    for coefficient in polynomial.into_iter() {
+        hash.update(coefficient.to_bytes());
+    }
+    hash.finalize().to_vec().to_base32()[0..4]
+        .try_into()
+        .expect("4 bech32 chars must fit 4 character arry")
+}
+
+/// Create a bech32m FROST backup
+///
+/// Requires that the threshold be no greater than 32.
+/// If using an integer (small) index, the backup will be 67 bech32 characters.
+/// If using a scalar index, the backup will be 118 bech32 characters.
+pub fn encode_backup<H: Default + Digest<OutputSize = U32>>(
+    threshold: usize,
+    polynomial: Vec<Point>,
+    secret_share: Scalar<Secret>,
+    share_index: BackupShareIndex,
+) -> Result<String, FrostBackupError> {
+    let mut data = [u5::default(); 1 + 4 + 52 + 52];
+
+    if threshold > 32 {
+        return Err(FrostBackupError::ThresholdTooBig);
+    } else if threshold == 0 {
+        return Err(FrostBackupError::ThresholdCantBeZero);
+    }
+    let threshold_u5 =
+        u5::try_from_u8((threshold - 1).to_le_bytes()[0]).map_err(FrostBackupError::Bech32Error)?;
+    data[0] = threshold_u5;
+
+    let polynomial_identifier_u5 = polynomial_identifier::<H>(polynomial);
+    for (i, byte) in polynomial_identifier_u5.into_iter().enumerate() {
+        data[1 + i] = byte;
+    }
+
+    let secret_share_u5 = secret_share.to_bytes().to_vec().to_base32();
```

Comment:
should we check the share is correct with respect to the polynomial?
Perhaps we should have an encode variant that does that.